### PR TITLE
Tpetra: Updates to deep_copy / fence counting tools

### PR DESCRIPTION
**Pull request by csiefer2**
Issue: #12106
Tests: New and shiny
Stakeholder feedback: Conversations with @cgcgcg 

The other tools are for timing, these are for counting.  Like in Regression testing.

### CODE REVIEW

File: packages/tpetra/core/src/Tpetra_Details_KokkosCounter.cpp [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+          count_same++;        
+      }
+    }
+
+  }// end DeepCopyCounterDetails
+
+
+  void DeepCopyCounter::start() {
+    DeepCopyCounterDetails::count_active=true;
+    Kokkos::Tools::Experimental::set_begin_deep_copy_callback(DeepCopyCounterDetails::kokkosp_begin_deep_copy);
```
Comment (jhux2) [inline packages/tpetra/core/src/Tpetra_Details_KokkosCounter.cpp:76]: I don't see see this stub in Kokkos tools?

Comment (jhux2) [reply packages/tpetra/core/src/Tpetra_Details_KokkosCounter.cpp:76]: Oh, it's in Kokkos proper

**Review by jhux2 — APPROVED** [review]
lgtm

**Comment by sebrowne** [comment]
Autotester is down (CEE GitLab is down and it needs that).

#12305

**Review by jhux2 — APPROVED** [review]
lgtm ;

**Review by jhux2 — APPROVED** [review]
lgtm

**Pull request merged**